[Feature] Ignore overridden method declaration

**Is your feature request related to a problem? Please describe.**
Currently, there is no clear way to override a method by declaring the `@Override' annotation.
For example:
```java
    @ConfigInterface
    public interface UserConfig extends User {

        @Override
        @Transient
        default String metadata(int index) {
            return metadata().get(index);
        }

    }
    
    public interface User {

        List<String> metadata();

        String metadata(int index);

    }
```
And if we will try to create `UserConfig` from valid configuration, we will see the exception (_see additional context_)

**Describe the solution you'd like**
When we will collect methods from class, we should avoid duplicate methods: overridden and their declaration. We could achieve this by including only overridden methods.

**Describe alternatives you've considered**
Currently none.

**Additional context**
```java
Exception in thread "main" java.lang.RuntimeException: org.spongepowered.configurate.serialize.SerializationException: [] of type com.bivashy.test.configurate.ProximateConfigurateTestInheritance$User: Interface methods should not have parameters: 'public abstract java.lang.String com.bivashy.test.configurate.ProximateConfigurateTestInheritance$User.test(int)'
	at com.bivashy.configurate.objectmapping.common.InterfaceMethodDiscoverer.lambda$shouldBeIgnored$6(InterfaceMethodDiscoverer.java:105)
	at java.base/java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
	at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1602)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
	at java.base/java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.anyMatch(ReferencePipeline.java:632)
	at com.bivashy.configurate.objectmapping.common.InterfaceMethodDiscoverer.shouldBeIgnored(InterfaceMethodDiscoverer.java:101)
	at com.bivashy.configurate.objectmapping.common.InterfaceMethodDiscoverer.collectMethods(InterfaceMethodDiscoverer.java:83)
	at com.bivashy.configurate.objectmapping.common.InterfaceMethodDiscoverer.discover(InterfaceMethodDiscoverer.java:73)
	at org.spongepowered.configurate.objectmapping.ObjectMapperFactoryImpl.newMapper(ObjectMapperFactoryImpl.java:131)
	at org.spongepowered.configurate.objectmapping.ObjectMapperFactoryImpl.computeMapper(ObjectMapperFactoryImpl.java:120)
	at org.spongepowered.configurate.objectmapping.ObjectMapperFactoryImpl.lambda$computeFromMap$5(ObjectMapperFactoryImpl.java:290)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1220)
	at org.spongepowered.configurate.objectmapping.ObjectMapperFactoryImpl.computeFromMap(ObjectMapperFactoryImpl.java:288)
	at org.spongepowered.configurate.objectmapping.ObjectMapperFactoryImpl.get(ObjectMapperFactoryImpl.java:109)
	at com.bivashy.configurate.objectmapping.common.InterfaceObjectMapperFactory.get(InterfaceObjectMapperFactory.java:65)
	at com.bivashy.configurate.objectmapping.common.InterfaceObjectMapperFactory.deserialize(InterfaceObjectMapperFactory.java:76)
	at org.spongepowered.configurate.AbstractConfigurationNode.get(AbstractConfigurationNode.java:151)
	at org.spongepowered.configurate.ConfigurationNode.get(ConfigurationNode.java:433)
Caused by: org.spongepowered.configurate.serialize.SerializationException: [] of type com.bivashy.test.configurate.ProximateConfigurateTestInheritance$User: Interface methods should not have parameters: 'public abstract java.lang.String com.bivashy.test.configurate.ProximateConfigurateTestInheritance$User.test(int)'
	at com.bivashy.configurate.objectmapping.common.InterfaceMethodDiscoverer.lambda$defaultBuilder$3(InterfaceMethodDiscoverer.java:53)
	at com.bivashy.configurate.objectmapping.common.InterfaceMethodDiscoverer.lambda$shouldBeIgnored$6(InterfaceMethodDiscoverer.java:103)
	... 24 more
```

--- a/common/src/main/java/com/bivashy/configurate/objectmapping/common/InterfaceMethodDiscoverer.java
+++ b/common/src/main/java/com/bivashy/configurate/objectmapping/common/InterfaceMethodDiscoverer.java
@@ -85,9 +85,10 @@ final class InterfaceMethodDiscoverer implements FieldDiscoverer<Map<String, Obj
             final String name = method.getName();
 
             MethodAdapter adapter = new MethodAdapter(method, type.getType());
-            if (duplicateMethods.contains(adapter))
+
+            boolean added = duplicateMethods.add(adapter);
+            if (!added)
                 continue;
-            duplicateMethods.add(adapter);
             if (shouldBeIgnored(method, type))
                 continue;
 

--- a/common/src/main/java/com/bivashy/configurate/objectmapping/common/InterfaceMethodDiscoverer.java
+++ b/common/src/main/java/com/bivashy/configurate/objectmapping/common/InterfaceMethodDiscoverer.java
@@ -7,8 +7,10 @@ import java.lang.reflect.Modifier;
 import java.lang.reflect.Type;
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
@@ -68,18 +70,24 @@ final class InterfaceMethodDiscoverer implements FieldDiscoverer<Map<String, Obj
         if (!clazz.isInterface())
             return null;
 
-        collectMethods(target, collector);
+        // duplicateMethods prevents processing of overridden methods
+        Set<MethodAdapter> duplicateMethods = new HashSet<>();
+        collectMethods(target, collector, duplicateMethods);
         for (AnnotatedType superType : getInterfaces(target, clazz)) {
-            collectMethods(superType, collector);
+            collectMethods(superType, collector, duplicateMethods);
         }
 
         return new ProxyInstanceFactory(clazz, invokers);
     }
 
-    private <V> void collectMethods(AnnotatedType type, final FieldCollector<Map<String, Object>, V> collector) {
+    private <V> void collectMethods(AnnotatedType type, final FieldCollector<Map<String, Object>, V> collector, Set<MethodAdapter> duplicateMethods) {
         for (Method method : GenericTypeReflector.erase(type.getType()).getDeclaredMethods()) {
             final String name = method.getName();
 
+            MethodAdapter adapter = new MethodAdapter(method, type.getType());
+            if (duplicateMethods.contains(adapter))
+                continue;
+            duplicateMethods.add(adapter);
             if (shouldBeIgnored(method, type))
                 continue;
 

--- a/common/src/main/java/com/bivashy/configurate/objectmapping/common/MethodAdapter.java
+++ b/common/src/main/java/com/bivashy/configurate/objectmapping/common/MethodAdapter.java
@@ -1,0 +1,42 @@
+package com.bivashy.configurate.objectmapping.common;
+
+import io.leangen.geantyref.GenericTypeReflector;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * MethodAdapter wraps a Method to enable comparisons and collections
+ * based on method signature (name, return type, parameters), ignoring the declaring class.
+ */
+final class MethodAdapter {
+
+    private final String name;
+    private final Type returnType;
+    private final Type[] parameterTypes;
+
+    public MethodAdapter(Method method, Type declaringType) {
+        this.name = method.getName();
+        this.returnType = GenericTypeReflector.getReturnType(method, declaringType);
+        this.parameterTypes = GenericTypeReflector.getParameterTypes(method, declaringType);
+    }
+
+    @Override
+    public boolean equals(Object object) {
+        if (this == object)
+            return true;
+        if (!(object instanceof MethodAdapter))
+            return false;
+        MethodAdapter adapter = (MethodAdapter) object;
+        return Objects.equals(name, adapter.name) && Objects.equals(returnType, adapter.returnType) &&
+                Objects.deepEquals(parameterTypes, adapter.parameterTypes);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(name, returnType, Arrays.hashCode(parameterTypes));
+    }
+
+}

--- a/common/src/test/java/com/bivashy/configurate/MethodOverrideTest.java
+++ b/common/src/test/java/com/bivashy/configurate/MethodOverrideTest.java
@@ -1,0 +1,116 @@
+package com.bivashy.configurate;
+
+import com.bivashy.configurate.objectmapping.ConfigInterface;
+import com.bivashy.configurate.objectmapping.common.InterfaceObjectMapperFactory;
+import com.bivashy.configurate.objectmapping.meta.Transient;
+import org.junit.jupiter.api.*;
+import org.spongepowered.configurate.ConfigurateException;
+import org.spongepowered.configurate.ConfigurationNode;
+import org.spongepowered.configurate.hocon.HoconConfigurationLoader;
+import org.spongepowered.configurate.serialize.SerializationException;
+import org.spongepowered.configurate.serialize.TypeSerializerCollection;
+
+import java.io.BufferedReader;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class MethodOverrideTest {
+
+    private final ConfigurationNode node = loadConfiguration();
+
+    static ConfigurationNode loadConfiguration() {
+        InputStream resourceStream = ObjectContractTest.class.getResourceAsStream("/inheritance.conf");
+        assertNotNull(resourceStream);
+        try {
+            return HoconConfigurationLoader.builder()
+                    .source(() -> new BufferedReader(new InputStreamReader(resourceStream)))
+                    .defaultOptions(opt ->
+                            opt.serializers(builder -> builder
+                                    .registerAll(TypeSerializerCollection.defaults())
+                                    .register(InterfaceObjectMapperFactory::applicable, new InterfaceObjectMapperFactory())
+                            ))
+                    .build()
+                    .load();
+        } catch (ConfigurateException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Test
+    void testSimpleOverride() throws SerializationException {
+        AnimalConfig animalConfig = node.node("animal").get(AnimalConfig.class);
+        assertNotNull(animalConfig, "animalConfig");
+
+        assertEquals(1, animalConfig.age());
+
+        assertEquals(animalConfig.age(), animalConfig.getAge());
+    }
+
+    @Test
+    void testGenericOverride() throws SerializationException {
+        AgedAnimalConfig agedAnimalConfig = node.node("animal").get(AgedAnimalConfig.class);
+        assertNotNull(agedAnimalConfig, "agedAnimalConfig");
+
+        assertEquals(1, agedAnimalConfig.age());
+        assertEquals(agedAnimalConfig.age(), agedAnimalConfig.getAge());
+        assertEquals(agedAnimalConfig.possibleColors(), agedAnimalConfig.getPossibleColors());
+        assertEquals(agedAnimalConfig.getColor(0), agedAnimalConfig.getPossibleColors().get(0));
+    }
+
+    public interface Animal {
+
+        int getAge();
+
+    }
+    @ConfigInterface
+    public interface AnimalConfig extends Animal {
+
+        @Transient
+        @Override
+        default int getAge() {
+            return age();
+        }
+
+        int age();
+
+    }
+    public interface Aged<T extends Number, I extends Number> {
+
+        T getAge();
+
+        String getColor(I index);
+
+        List<String> getPossibleColors();
+
+    }
+    @ConfigInterface
+    public interface AgedAnimalConfig extends Aged<Integer, Integer> {
+
+        @Transient
+        @Override
+        default Integer getAge() {
+            return age();
+        }
+
+        @Transient
+        @Override
+        default String getColor(Integer index) {
+            return possibleColors().get(index);
+        }
+
+        @Transient
+        @Override
+        default List<String> getPossibleColors() {
+            return possibleColors();
+        }
+
+        List<String> possibleColors();
+
+        int age();
+
+    }
+
+}

--- a/common/src/test/resources/inheritance.conf
+++ b/common/src/test/resources/inheritance.conf
@@ -1,6 +1,7 @@
 {
   animal: {
-    age: 1
+    age: 1,
+    possible-colors: ["white", "black"]
   },
   dog: {
     age: 2

